components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/logout/idpinit/processor/FederatedIdpInitLogoutProcessor.java

```diff
@@ -0,0 +1,562 @@
+
+import static org.wso2.carbon.identity.application.authenticator.oidc.OIDCAuthenticatorConstants.AUTHENTICATOR_NAME;
+import static org.wso2.carbon.identity.application.authenticator.oidc.OIDCAuthenticatorConstants.OIDC_BACKCHANNEL_LOGOUT_ENDPOINT_URL_PATTERN;
+import static org.wso2.carbon.identity.application.authenticator.oidc.util.OIDCErrorConstants.ErrorMessages;
+
+/**
+ * Processes the OIDC federated idp initiated logout requests.
+ */
+public class FederatedIdpInitLogoutProcessor extends IdentityProcessor {
+
+    private static final Log log = LogFactory.getLog(FederatedIdpInitLogoutProcessor.class);
+
+    @Override
+    public IdentityResponse.IdentityResponseBuilder process(IdentityRequest identityRequest) throws FrameworkException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Request processing started by OIDC FederatedIdpInitLogoutProcessor.");
+        }
+        return handleOIDCFederatedLogoutRequest(identityRequest);
+    }
+
+    /**
+     * Handles the logout request according to OIDC Back-channel logout specification.
+     *
+     * @param logoutRequest
+     * @return IdentityResponse.IdentityResponseBuilder.
+     * @throws LogoutClientException Exception occurred due to error in the logout request.
+     * @throws LogoutServerException Exception occurred from IS.
+     */
+    protected IdentityResponse.IdentityResponseBuilder handleOIDCFederatedLogoutRequest(
+            IdentityRequest logoutRequest) throws LogoutClientException, LogoutServerException {
+
+        String sub = "";
+        try {
+            String logoutToken = logoutRequest.getParameter(OIDCAuthenticatorConstants.LOGOUT_TOKEN);
+            if (StringUtils.isBlank(logoutToken)) {
+                throw new LogoutClientException(ErrorMessages.LOGOUT_TOKEN_EMPTY_OR_NULL.getCode(),
+                        ErrorMessages.LOGOUT_TOKEN_EMPTY_OR_NULL.getMessage());
+            }
+            if (log.isDebugEnabled()) {
+                log.debug("Started handling the federated IdP Initiated Logout request. Logout Token: " + logoutToken);
+            }
+
+            SignedJWT signedJWT = SignedJWT.parse(logoutToken);
+            JWTClaimsSet claimsSet = signedJWT.getJWTClaimsSet();
+            // Get the identity provider for the issuer of the logout token.
+            String tenantDomain = logoutRequest.getTenantDomain();
+            IdentityProvider identityProvider =
+                    getIdentityProvider(claimsSet.getIssuer(), tenantDomain);
+            validateLogoutToken(signedJWT, identityProvider);
+            sub = claimsSet.getSubject();
+            // Retrieve the federated user id from the IDN_AUTH_USER table.
+            int tenantId = tenantDomain == null ? -1 : IdentityTenantUtil.getTenantId(tenantDomain);
+            String userId = UserSessionStore.getInstance()
+                    .getFederatedUserId(sub, tenantId, Integer.parseInt(identityProvider.getId()));
```

Comment:
Move this to a private method which returns the userId when you provide sub, tenantDomain, IDP